### ISSUE TITLE
Unable to program a M27C801 1MB EPROM

### ISSUE THREAD

**Issue by DavidGriffith** [DavidGriffith/minipro-import-test#139]
In GitLab by @rchoudhary1 on May 8, 2019, 02:51

I bought some M27C801 1MB EPROM ICs and attempted to program them with minipro.

First, I created a file called `mem_blank.hex` which was an 1MB file that contained all 0xFF's. This corresponds to a fully erased EPROM. 

I put the chip in a UV eraser and the read the contents with `minipro -p "M27C801" -r mem_read.hex`. I then used the command `cmp mem_read.hex mem_blank.hex` to test if the chip was indeed fully erased. It passed that test.

Then I generated a file called `mem_write.hex`. It basically was an 1MB file filled with random bytes. I wrote this to the IC using `minipro -p "M27C801" -w mem_write.hex`. 

The result I got from minipro was:

```
Found TL866II+ 04.2.86 (0x256)
Warning: Firmware is out of date.
  Expected  04.2.99 (0x263)
  Found     04.2.86 (0x256)
Chip ID OK: 0x2042
Erasing... 0.25Sec OK
Writing Code...  292.90Sec  OK
Reading Code...  9.09Sec  OK
Verification failed at address 0x5B34: File=0x5E, Device=0x00
```

To confirm for myself, I read the contents again using `minipro -p "M27C801" -r mem_read.hex` and ran `wdiff <(xxd mem_read.hex) <(xxd mem_write.hex) | grep -E "\[\-.+\{\+"` to see the differences word by word. This is an excerpt of the results:

```
00005b30: f62e 8bdf [-0004-] {+5e04+} d981 adf1 6025 9e70 b6db  [-..........`%.p..-]  {+....^.....`%.p..+}
00009580: 3d1b d228 [-4100-] {+4173+} 132b b3e4 1f17 6e91 53a3  [-=..(A..+....n.S.-]  {+=..(As.+....n.S.+}
00009790: [-c600-] {+c6bf+} 97bb a72a 9bb4 94a3 f91c 6372 6993  .....*......cri.
0000ba80: eb75 [-0002-] {+6e02+} 89c4 15a4 774f d1a4 dff6 102f  [-.u......wO...../-]  {+.un.....wO...../+}
...
000fdf70: 96b6 8374 8125 dda6 73f7 [-0004-] {+b904+} 2280 119c  ...t.%..s..."...
000ffe70: 3694 6748 1b89 de4b a5a1 b244 3118 [-0045  6.gH...K...D1..E-] {+4545  6.gH...K...D1.EE+}
```

I'm not sure why the write operations aren't working. Does it have to do with power supply to the IC? Or could it be because of my out of date firmware? 

The thing about upgrading is that I'm not sure how to upgrade to version 04.2.99 _specifically_. I fear that I will upgrade to a much newer version which is also incompatible with minipro and I won't be able to roll back.

**Comment by DavidGriffith** [comment]
In GitLab by @rchoudhary1 on May 8, 2019, 23:06

**UPDATE:** So I downloaded the Xgpro software on a Windows partition and updated the firmware. Then I wrote `mem_write.hex` to the IC using that. After that I performed the "verify chip with buffer" action and it passed. 

I then switched back to my Mac. First I checked the minipro version:

```
(base) Desktop $ minipro -t
Found TL866II+ 04.2.103 (0x267)
Warning: Firmware is newer than expected.
  Expected  04.2.99 (0x263)
  Found     04.2.103 (0x267)
TL866II+: hardware_check not implemented
```

I then read the IC contents with `minipro -p "M27C801" -r mem_read.hex`. When I ran `cmp mem_read.hex mem_write.hex`, it said the files were identical.

After that I went and erased the chip. I confirmed that it was blank using both Xgpro and minipro. 

Then, I re-generated `mem_write.hex` to contain new random bytes. I then ran `minipro -p "M27C801" -w mem_write.hex` to write to the freshly erased IC.

Again I read the IC contents with `minipro -p "M27C801" -r mem_read.hex` and then ran `cmp mem_read.hex mem_write.hex`, and it said the files were identical.

So it looks like the issue was outdated firmware.
